src/core/a-entity.js

```diff
@@ -461,37 +495,60 @@ var proto = Object.create(ANode.prototype, {
    */
   setAttribute: {
     value: function (attr, value, componentPropValue) {
+      var debug = this.sceneEl && this.sceneEl.getAttribute('debug');
       var self = this;
       var component = this.components[attr] || components[attr];
-      var partialComponentData;
-      value = value === undefined ? '' : value;
       var componentObj = value;  // Deserialized value to send to the component.
-      var componentStr = value;  // Serialized value to send to the DOM.
-      var oldValue;
+      // if there's a previous value it will be cached in the component.
+      var oldValue = this.components[attr] && this.components[attr].attrValue;
 
       if (component) {
         if (typeof value === 'string' && componentPropValue !== undefined) {
           // Update currently-defined component data with the new property value.
-          // Use native setAttribute in order not to double-parse properties.
-          partialComponentData = styleParser.parse(
-            HTMLElement.prototype.getAttribute.call(this, attr)) || {};
-          partialComponentData[value] = componentPropValue;
-          componentObj = partialComponentData;
+          // Use cached attributes to prevent unnecessary string parsings.
+          oldValue = oldValue ? utils.extend({}, oldValue) : {};
+          componentObj = component.attrValue || {};
+          componentObj[value] = componentPropValue;
         }
-        componentStr = component.stringify(componentObj);
+        self.setEntityAttribute(attr, oldValue, componentObj);
+        // On debug mode we write the component state to the DOM attributes
+        if (debug) { component.flushToDOM(); }
+        return;
       }
-      oldValue = this.getAttribute(attr);
-      ANode.prototype.setAttribute.call(self, attr, componentStr);
-      self.setEntityAttribute(attr, oldValue, componentObj);
+
+      ANode.prototype.setAttribute.call(self, attr, value);
+      if (attr === 'mixin') { self.setEntityAttribute('mixin', undefined, componentObj); }
     },
     writable: window.debug
   },
 
   /**
-   * If `attr` is a component, returns JUST the component data specified in the HTML
-   * by parsing the style-like string into an object. Like a partial version of
-   * `getComputedAttribute` as returned component data does not include applied mixins or
-   * defaults.
+   * To make the DOM attributes reflect the state of the components.
+   *
+   * @param {bool} recursive - It calls updateDOM on the children
+   **/
+  flushToDOM: {
+    value: function (recursive) {
+      var components = this.components;
+      var children = this.children;
+      var child;
+      var i;
+      Object.keys(components).forEach(function updateAttribute (name) {
```

Comment:
OK. got confuse since we use `update` and overload `attribute` a lot. might be good to reuse the word `flush`